### PR TITLE
OpenAPI 3.0

### PR DESCRIPTION
**Pull request by raymondfeng**
### Description

Upgrade https://github.com/strongloop/loopback-connector-swagger to support OpenAPI 3.0.

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

**Comment by raymondfeng** [comment]
Duplicate of #1

**Comment by bajtos** [comment]
The patch is pretty large (`+1,900 −1,271` lines of code), I don't have bandwidth and energy to review it in detail again. I already posted quite a few comments in https://github.com/strongloop/loopback-connector-swagger/pull/52, I think many of them applies to this pull request too.

My main concern is about the inclusion of an LB4 example app as a standalone project inside this repository. How are you integrating `npm install`, `npm run build` and possibly also `npm test` into the CI build pipeline of the connector project?

IMO, the LB4 example app should live in `loopback-next` monorepo.

**Comment by raymondfeng** [comment]
The main purpose of the embedded LB4 app is for testing OpenAPI 3.0 compliant apis. I already make it part of the test as follows:

1. In https://github.com/strongloop/loopback-connector-openapi/blob/openapi-3.0/package.json#L10, there is a pretest script to build the LB4 app.

2. The LB4 app is programmatically started/stopped as part of the test - https://github.com/strongloop/loopback-connector-openapi/blob/openapi-3.0/test/test-connector-openapi3.js#L177. There is no extra CI logistics involved.

**Comment by raymondfeng** [comment]
@bajtos I switched to `@loopback/example-todo` as a dev dependency for testing.

File: test/test-connector-auth.js [hunk 1]
```diff
@@ -1,107 +1,119 @@
...
 // This file is licensed under the MIT License.
 // License text available at https://opensource.org/licenses/MIT
 
 'use strict';
 
-var assert = require('assert');
-var should = require('should');
-var loopback = require('loopback');
+// eslint-disable-next-line no-unused
+/* eslint-disable camelcase */
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply test/test-connector-auth.js]: Yes. `should` is unused and there are a few non-camelcase keys.

File: test/test-connector-auth.js [hunk 2]
```diff
@@ -1,107 +1,119 @@
...
-      });
+      givenDataSource(
+        'test/fixtures/2.0/petstore.json',
+        {
+          basic: {
+            username: 'aaabbbccc',
+            password: 'header',
+          },
+        },
+        (err, req) => {
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply test/test-connector-auth.js]: Some of the tests still use callback style to handle hooks.

**Comment by jannyHou** [comment]
Great to see the upgrade! I may not have time to review it this week(no need wait for my approve if it's ready to land), will take a look in April.

File: test/test-connector-auth.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (bajtos) [reply test/test-connector-auth.js]: Well if `should` is not used, then why are we keeping it around? Can you simply remove L10?

File: test/test-connector-auth.js [hunk 3]
```diff
@@ -102,7 +107,7 @@ describe('OpenAPI connector - security', function() {
   });
 });
 
-function givenDataSource(spec, authz, done) {
+async function givenDataSource(spec, authz, done) {
```
Comment (bajtos) [inline test/test-connector-auth.js]: A callback-based function returning a promise? That looks suspicious to me. What are you trying to achieve by this change to `async function`? Did you perhaps want to rewrite this helper in a similar style as you did in `test/test-connector-swagger2.js`?

File: test/test-connector-auth.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (raymondfeng) [reply test/test-connector-auth.js]: I believe `should` is required so that `*.should` works.

File: test/test-connector-auth.js (hunk 3 above)
Comment (raymondfeng) [reply test/test-connector-auth.js]: Fixed.

File: lib/spec-resolver.js [hunk 4]
```diff
@@ -1,74 +1,22 @@
...
 
-      default:
-        cb(new Error('Invalid specification file type.'), null);
-    }
-  } else {
-    process.nextTick(function() {
-      cb(new Error('Invalid swagger specification type'), null);
-    });
+exports.resolve = function resolveSpec(spec, options, cb) {
+  if (typeof options === 'function' && !cb) {
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply lib/spec-resolver.js]: The old logic is already supported by `swagger-parser` module.

File: test/test-connector-auth.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (bajtos) [reply test/test-connector-auth.js]: Makes sense 👍

**Pull request merged**